### PR TITLE
Data 194 kin3 change output

### PR DESCRIPTION
**Pull request by oren-a**
* Changing the output schema of both S3 and Postgres, to include more informations per operations.

Configuration:
* Added retry delay config
* Marked many of the logging information as 'debug' rather than info.

Postgres:
* Removed creations/payments tables and unifying them under 'operations' table, using a 'type' column to differ between them. We will also support 'merge account' type.
* Added indexing and constraints to the postgres tables

S3:
* Fixed a bug caused by the delimiter character: ',' appearing in the memo. Columns that may have it, should be escaped according to the csv standards (put the column in double-quotes and each double-quote inside it should be appended with an additional double-quote as well)

Bug fixes:
* when the output adapter fails, re-initialize it

### CODE REVIEW

File: python/adapters/hc_storage_adapter.py [hunk 1]
```diff
@@ -68,32 +60,19 @@ def payments_output_schema():
...
-        return {
-            'source': str,
-            'destination': str,
-            'starting_balance': int,
-            'memo': str,
-            'tx_fee': int,
-            'tx_charged_fee': int,
-            'op_index': int,
-            'tx_status': str,
+            'op_order': int,
```
Comment (Ronserruya) [inline python/adapters/hc_storage_adapter.py]: does this mean the order of the operation in the tx? if so, op index would be the correct term used in the blockchain

File: python/adapters/postgres_storage_adapter.py [hunk 2]
```diff
@@ -53,78 +43,57 @@ def _commit(self):
...
-        creation['op_status'] = op_status
-        creation['hash'] = tx_hash
-        creation['time'] = datetime.utcfromtimestamp(timestamp)
-
-        return creation
+    def convert_operation(self, source, destination, amount, tx_order, tx_memo, tx_account, tx_account_sequence,
+                          tx_fee, tx_charged_fee, tx_status, tx_hash, op_order, op_status, op_type, timestamp,
+                          is_signed_by_app, ledger_file_name, ledger_sequence):
+
+        operation = dict.fromkeys(self.operation_output_schema())
```
Comment (Ronserruya) [inline python/adapters/postgres_storage_adapter.py:50]: why is this needed if you override everything?

File: python/adapters/postgres_storage_adapter.py [hunk 3]
```diff
@@ -53,78 +43,57 @@ def _commit(self):
...
+        operation['tx_status'] = tx_status
+        operation['tx_hash'] = tx_hash
+        operation['op_order'] = op_order
+        operation['op_status'] = op_status
+        operation['op_type'] = op_type
+        operation['timestamp'] = datetime.utcfromtimestamp(timestamp)
+        operation['is_signed_by_app'] = is_signed_by_app
+        operation['ledger_file_name'] = ledger_file_name
+        operation['ledger_sequence'] = ledger_sequence
+
```
Comment (Ronserruya) [inline python/adapters/postgres_storage_adapter.py:69]: Would this work for you?

```python
params = list(inspect.signature(self.convert_operation).parameters.keys())
loc = locals()
operation = {var: loc[var] for var in params}
operation['timestamp'] = datetime.utcfromtimestamp(timestamp)
```

Smaller and more flexible if you want to add more variables

File: python/adapters/postgres_storage_adapter.py [hunk 4]
```diff
@@ -53,78 +43,57 @@ def _commit(self):
...
         """
-        :return: A dictionary of columns saved by the History collector.
+        :return: A dictionary of columns saved by the History collector.\
           Key - name, Value - string literal of postgres type
         """
 
         return {
-            'source': 'varchar(56) not NULL',
-            'destination': 'varchar(56) not NULL',
+            'source': 'varchar(64) not NULL',
```
Comment (Ronserruya) [inline python/adapters/postgres_storage_adapter.py]: why 64? source and destination can only be 56

File: python/adapters/postgres_storage_adapter.py [hunk 5]
```diff
@@ -53,78 +43,57 @@ def _commit(self):
...
             'amount': 'BIGINT not NULL',
-            'memo_text': 'varchar(28)',
-            'fee': 'INT not NULL',
-            'fee_charged': 'INT not NULL',
-            'operation_index': 'INT not NULL',
-            'tx_status': 'text',
-            'op_status': 'text',
-            'hash': 'varchar(64) not NULL',
-            'time': 'TIMESTAMP not NULL'
+            'tx_order': 'INT not NULL',
```
Comment (Ronserruya) [inline python/adapters/postgres_storage_adapter.py:83]: tx_index would be more correct, but not a must

File: python/adapters/s3_storage_adapter.py [hunk 6]
```diff
@@ -107,64 +99,38 @@ def _rollback(self):
...
-        :return: A dictionary of columns saved by the History collector. Key - name, Value - type
-        """
-
-        schema = HistoryCollectorStorageAdapter.creations_output_schema()
-        schema.update({'type': str})
-        return schema
-
-    def __init_operations_to_save(self):
-        self.operations_to_save = []
+    def convert_operation(self, source, destination, amount, tx_order, tx_memo, tx_account, tx_account_sequence,
```
Comment (Ronserruya) [inline python/adapters/s3_storage_adapter.py:102]: the same thing with the previous comment can simplify this

File: python/adapters/s3_storage_adapter.py [hunk 7]
```diff
@@ -177,11 +143,11 @@ def __test_connection(self):
             self.get_last_file_sequence()
             # Trying to write 'test' ledger, then delete it. Not using the 'save' method, as it will rewrite last_file
             self.file_name = 'test'
-            self._save_creations([{'source': 'GCQTAWULBNFLBAEQLEN6FDGGCPYTVZ3Y55AB4F7HSTMQKNX3HZINMQJM',
-                                   'destination': 'GDDFYG3OSTSHADS7SP6TZ4XM62EQ522CI7UYJSNAETGJJCGOX66TP5Q5',
+            self._save_operations([{'source': 'JUST_A_TEST',
```
Comment (Ronserruya) [inline python/adapters/s3_storage_adapter.py:146]: doesn't really matter, but I would use a real address instead of some string, cant know what absurd bug can show up

File: python/enums.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+from enum import Enum, unique
+
+
+@unique
+class OperationType(Enum):
```
Comment (Ronserruya) [inline python/enums.py:5]: Already exists in the sdk

```
In [4]: kin_base.stellarxdr.StellarXDR_const.OperationType                                                                                                                   
Out[4]: 
{0: 'CREATE_ACCOUNT',
 1: 'PAYMENT',
 2: 'PATH_PAYMENT',
 3: 'MANAGE_OFFER',
 4: 'CREATE_PASSIVE_OFFER',
 5: 'SET_OPTIONS',
 6: 'CHANGE_TRUST',
 7: 'ALLOW_TRUST',
 8: 'ACCOUNT_MERGE',
 9: 'INFLATION',
 10: 'MANAGE_DATA',
 11: 'BUMP_SEQUENCE'}

```

File: python/adapters/s3_storage_adapter.py [hunk 9]
```diff
@@ -107,64 +99,38 @@ def _rollback(self):
...
+        operation['op_status'] = op_status
+        operation['op_type'] = op_type
+        operation['timestamp'] = datetime.utcfromtimestamp(timestamp)
+        operation['is_signed_by_app'] = is_signed_by_app
+        operation['ledger_file_name'] = ledger_file_name
+        operation['ledger_sequence'] = ledger_sequence
+
+        # Fields that could possibly contain the separator character, should be enclosed in double-quotes, and each
+        #  double-quote inside should have a preceding double-quote
+        fields_to_format = ['tx_memo']
```
Comment (Ronserruya) [inline python/adapters/s3_storage_adapter.py:128]: pandas ```to_csv``` has many args that might be better for this, 

```
quoting
quotechar
doublequote
escapechar
```

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html

**Comment by Ronserruya** [comment]
Overall looks good, left some comments :)

I suggest you try to read a bit on [dataclasses](https://docs.python.org/3/library/dataclasses.html) and the [pydantic](https://github.com/samuelcolvin/pydantic) package, If I knew about them when I originally wrote it, I would have used them from the start to define operations.

**Review by Ronserruya — APPROVED** [review]

File: python/adapters/hc_storage_adapter.py (hunk 1 above)
Comment (oren-a) [reply python/adapters/hc_storage_adapter.py]: Thanks, I didn't know there's a specific term in blockchain for that. I changed it based on a design document I got for this change, but I agree - it will make more sense to keep that as 'index', I'll change it

File: python/adapters/postgres_storage_adapter.py (hunk 2 above)
Comment (oren-a) [reply python/adapters/postgres_storage_adapter.py:50]: To guarantee that all of the keys exists based on the output schema, even if they are empty, or any future assignment will be missed, but you next comment could prevent that
